executor/cluster_reader.go

```diff
@@ -215,3 +218,231 @@ func (e *clusterConfigRetriever) retrieve(ctx sessionctx.Context) ([][]types.Dat
+		}
+
+		// To avoid search log interface overload, the user should specify at least one pattern
+		// in normally SQL. (But in test mode we should relax this limitation)
+		if len(req.Patterns) == 0 && len(req.Levels) == 0 {
+			return errors.New("denied to scan full logs (use `SELECT * FROM cluster_log WHERE message LIKE '%'` explicitly if intentionally)")
+		}
+	}
+
+	var results []chan logStreamResult
+	for _, srv := range serversInfo {
+		typ := srv.ServerType
+		// Skip some node type which has been filtered in WHERE cluase
+		// e.g: SELECT * FROM cluster_log WHERE type='tikv'
+		if len(e.extractor.NodeTypes) > 0 && !e.extractor.NodeTypes.Exist(typ) {
+			continue
+		}
+		address := srv.Address
+		// Skip some node address which has been filtered in WHERE cluase
+		// e.g: SELECT * FROM cluster_log WHERE address='192.16.8.12:2379'
+		if len(e.extractor.Addresses) > 0 && !e.extractor.Addresses.Exist(address) {
+			continue
+		}
+		statusAddr := srv.StatusAddr
+		if len(statusAddr) == 0 {
+			ctx.GetSessionVars().StmtCtx.AppendWarning(errors.Errorf("%s node %s does not contain status address", typ, address))
+			continue
+		}
+		ch := make(chan logStreamResult)
+		results = append(results, ch)
+
+		go func(ch chan logStreamResult, serverType, address, statusAddr string) {
+			util.WithRecovery(func() {
+				defer close(ch)
+
+				// The TiDB provide diagnostics service via status address
+				remote := address
+				if serverType == "tidb" {
+					remote = statusAddr
+				}
+				conn, err := grpc.Dial(remote, opt)
+				if err != nil {
+					ch <- logStreamResult{addr: address, typ: serverType, err: err}
+					return
+				}
+				defer terror.Call(conn.Close)
+
+				cli := diagnosticspb.NewDiagnosticsClient(conn)
+				stream, err := cli.SearchLog(context.Background(), req)
+				if err != nil {
+					ch <- logStreamResult{addr: address, typ: serverType, err: err}
+					return
+				}
+
+				for {
+					res, err := stream.Recv()
+					if err != nil && err == io.EOF {
+						return
+					}
+					if err != nil {
+						ch <- logStreamResult{addr: address, typ: serverType, err: err}
+						return
+					}
+					ch <- logStreamResult{next: ch, addr: address, typ: serverType, messages: res.Messages}
+				}
+			}, nil)
+		}(ch, typ, address, statusAddr)
+	}
+
+	e.results = results
+	return nil
+}
+
+func (e *clusterLogRetriever) retrieve(ctx sessionctx.Context) ([][]types.Datum, error) {
+	if e.extractor.SkipRequest || e.isDrained {
+		return nil, nil
+	}
+
+	if !e.retrieving {
+		e.retrieving = true
+		if err := e.startRetrieving(ctx); err != nil {
+			e.isDrained = true
+			return nil, err
+		}
+
+		// initialize the heap
+		e.heap = &logResponseHeap{}
+		n := len(e.results)
+		for i := n - 1; i >= 0; i-- {
+			result := <-e.results[i]
+			if result.err != nil || len(result.messages) == 0 {
+				if result.err != nil {
+					ctx.GetSessionVars().StmtCtx.AppendWarning(result.err)
+				}
+				e.results = append(e.results[:i], e.results[i+1:]...)
```

Comment:
Is `e.results` needed any longer? If not, why modify it?

Reply:
nice catch.